File: Sources/HoverPocket/Capabilities/CapabilityCompatibility.swift
import Foundation

enum PocketCapabilityLifecycleStatus: String, Codable, Sendable {
    case active
    case deprecated
    case removed
}

struct PocketCapabilityLifecycleRecord: Equatable, Sendable {
    let key: PocketCapabilityKey
    let status: PocketCapabilityLifecycleStatus
    let introducedInHostVersion: String
    let deprecatedInHostVersion: String?
    let removalNotBeforeHostVersion: String?
    let replacement: PocketCapabilityKey?
    let migrationID: String?
    let noticeKey: String
}

struct PocketCapabilityReferenceMigration: Equatable, Sendable {
    let id: String
    let source: PocketCapabilityKey
    let target: PocketCapabilityKey
}

struct PocketCapabilityCompatibilityIssue: Equatable, Sendable {
    let key: PocketCapabilityKey
    let status: PocketCapabilityLifecycleStatus
    let replacement: PocketCapabilityKey
    let migrationID: String
    let removalNotBeforeHostVersion: String
    let noticeKey: String
}

enum PocketCapabilityCompatibilityError: Error, Equatable, Sendable {
    case invalidPolicy(String)
    case migrationUnavailable(PocketCapabilityKey)
}

struct PocketCapabilityCompatibilityCatalog: Sendable {
    static let currentHostVersion = "1.0.0"

    static let builtIn: PocketCapabilityCompatibilityCatalog = {
        do {
            return try PocketCapabilityCompatibilityCatalog(
                hostVersion: currentHostVersion,
                records: [],
                migrations: []
            )
        } catch {
            preconditionFailure("Invalid built-in capability compatibility catalog: \(error)")
        }
    }()

    let hostVersion: String
    private let records: [PocketCapabilityKey: PocketCapabilityLifecycleRecord]
    private let migrations: [PocketCapabilityKey: PocketCapabilityReferenceMigration]

    init(
        hostVersion: String,
        records: [PocketCapabilityLifecycleRecord],
        migrations: [PocketCapabilityReferenceMigration]
    ) throws {
        guard Self.semanticVersion(hostVersion) != nil else {
            throw PocketCapabilityCompatibilityError.invalidPolicy("host_version")
        }

        var recordMap: [PocketCapabilityKey: PocketCapabilityLifecycleRecord] = [:]
        for record in records {
            guard Self.validKey(record.key),
                  Self.semanticVersion(record.introducedInHostVersion) != nil,
                  recordMap[record.key] == nil else {
                throw PocketCapabilityCompatibilityError.invalidPolicy("record")
            }
            try Self.validate(record: record, hostVersion: hostVersion)
            recordMap[record.key] = record
        }

        var migrationMap: [PocketCapabilityKey: PocketCapabilityReferenceMigration] = [:]
        var migrationIDs: Set<String> = []
        for migration in migrations {
            guard Self.validMigrationID(migration.id),
                  Self.validKey(migration.source),
                  Self.validKey(migration.target),
                  migration.source != migration.target,
                  migrationMap[migration.source] == nil,
                  migrationIDs.insert(migration.id).inserted else {
                throw PocketCapabilityCompatibilityError.invalidPolicy("migration")
            }
            guard let record = recordMap[migration.source],
                  record.status != .active,
                  record.replacement == migration.target,
                  record.migrationID == migration.id else {
                throw PocketCapabilityCompatibilityError.invalidPolicy("migration_binding")
            }
            migrationMap[migration.source] = migration
        }

        for record in recordMap.values where record.status != .active {
            guard migrationMap[record.key] != nil else {
                throw PocketCapabilityCompatibilityError.invalidPolicy("migration_missing")
            }
            guard let replacement = record.replacement,
                  recordMap[replacement]?.status != .removed else {
                throw PocketCapabilityCompatibilityError.invalidPolicy("replacement_removed")
            }
        }

        for source in migrationMap.keys {
            var visited: Set<PocketCapabilityKey> = [source]
            var current = migrationMap[source]?.target
            while let key = current, let next = migrationMap[key]?.target {
                guard visited.insert(key).inserted else {
                    throw PocketCapabilityCompatibilityError.invalidPolicy("migration_cycle")
                }
                current = next
            }
        }

        self.hostVersion = hostVersion
        self.records = recordMap
        self.migrations = migrationMap
    }

    func status(for key: PocketCapabilityKey) -> PocketCapabilityLifecycleStatus {
        records[key]?.status ?? .active
    }

    func issue(for key: PocketCapabilityKey) -> PocketCapabilityCompatibilityIssue? {
        guard let record = records[key],
              record.status != .active,
              let replacement = record.replacement,
              let migrationID = record.migrationID,
              let removalNotBefore = record.removalNotBeforeHostVersion else {
            return nil
        }
        return PocketCapabilityCompatibilityIssue(
            key: key,
            status: record.status,
            replacement: replacement,
            migrationID: migrationID,
            removalNotBeforeHostVersion: removalNotBefore,
            noticeKey: record.noticeKey
        )
    }

    func migration(for key: PocketCapabilityKey) throws -> PocketCapabilityReferenceMigration {
        guard let migration = migrations[key] else {
            throw PocketCapabilityCompatibilityError.migrationUnavailable(key)
        }
        return migration
    }

    func requireRuntimeExecutable(_ key: PocketCapabilityKey) throws {
        guard let record = records[key], record.status == .removed else { return }
        throw CapabilityBrokerError.removedCapability(key, record.replacement)
    }

    private static func validate(record: PocketCapabilityLifecycleRecord, hostVersion: String) throws {
        guard validNoticeKey(record.noticeKey),
              let introduced = semanticVersion(record.introducedInHostVersion),
              let host = semanticVersion(hostVersion),
              compare(introduced, host) != .orderedDescending else {
            throw PocketCapabilityCompatibilityError.invalidPolicy("record_version")
        }

        switch record.status {
        case .active:
            guard record.deprecatedInHostVersion == nil,
                  record.removalNotBeforeHostVersion == nil,
                  record.replacement == nil,
                  record.migrationID == nil else {
                throw PocketCapabilityCompatibilityError.invalidPolicy("active_fields")
            }
        case .deprecated, .removed:
            guard let deprecated = record.deprecatedInHostVersion.flatMap(semanticVersion),
                  let removal = record.removalNotBeforeHostVersion.flatMap(semanticVersion),
                  let replacement = record.replacement,
                  validKey(replacement), replacement != record.key,
                  let migrationID = record.migrationID,
                  validMigrationID(migrationID),
                  compare(introduced, deprecated) != .orderedDescending,
                  compare(deprecated, host) != .orderedDescending,
                  compare(deprecated, removal) == .orderedAscending else {
                throw PocketCapabilityCompatibilityError.invalidPolicy("deprecation_window")
            }
            if record.status == .removed, compare(host, removal) == .orderedAscending {
                throw PocketCapabilityCompatibilityError.invalidPolicy("removed_too_early")
            }
        }
    }

    private static func validKey(_ key: PocketCapabilityKey) -> Bool {
        key.version >= 1
            && key.id.range(
                of: "^[a-z][a-z0-9]*(?:\\.[a-z][a-z0-9_]*){2,}$",
                options: .regularExpression
            ) != nil
    }

    private static func validMigrationID(_ value: String) -> Bool {
        value.unicodeScalars.count <= 128
            && value.range(
                of: "^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)+$",
                options: .regularExpression
            ) != nil
    }

    private static func validNoticeKey(_ value: String) -> Bool {
        value.unicodeScalars.count <= 160
            && value.range(
                of: "^[a-z][a-z0-9]*(?:\\.[a-z0-9_-]+)+$",
                options: .regularExpression
            ) != nil
    }

    private static func semanticVersion(_ value: String) -> [Int]? {
        guard value.unicodeScalars.count <= 64,
              let match = value.wholeMatch(of: /^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)$/),
              let major = Int(match.1),
              let minor = Int(match.2),
              let patch = Int(match.3) else {
            return nil
        }
        return [major, minor, patch]
    }

    private static func compare(_ lhs: [Int], _ rhs: [Int]) -> ComparisonResult {
        for (left, right) in zip(lhs, rhs) where left != right {
            return left < right ? .orderedAscending : .orderedDescending
        }
        return .orderedSame
    }
}

File: windows/src/HoverPocket.Shell/Capabilities/CapabilityCompatibility.cs
using System.Text.RegularExpressions;

namespace HoverPocket.Shell.Capabilities;

internal enum PocketCapabilityLifecycleStatus
{
    Active,
    Deprecated,
    Removed
}

internal sealed record PocketCapabilityLifecycleRecord(
    PocketCapabilityKey Key,
    PocketCapabilityLifecycleStatus Status,
    string IntroducedInHostVersion,
    string? DeprecatedInHostVersion,
    string? RemovalNotBeforeHostVersion,
    PocketCapabilityKey? Replacement,
    string? MigrationId,
    string NoticeKey);

internal sealed record PocketCapabilityReferenceMigration(
    string Id,
    PocketCapabilityKey Source,
    PocketCapabilityKey Target);

internal sealed record PocketCapabilityCompatibilityIssue(
    PocketCapabilityKey Key,
    PocketCapabilityLifecycleStatus Status,
    PocketCapabilityKey Replacement,
    string MigrationId,
    string RemovalNotBeforeHostVersion,
    string NoticeKey);

internal sealed class PocketCapabilityCompatibilityException(string code) : Exception(code)
{
    public string Code { get; } = code;
}

internal sealed class PocketCapabilityCompatibilityCatalog
{
    internal const string CurrentHostVersion = "1.0.0";

    private static readonly Regex CapabilityIdPattern = new(
        "^[a-z][a-z0-9]*(?:\\.[a-z][a-z0-9_]*){2,}$",
        RegexOptions.CultureInvariant | RegexOptions.NonBacktracking);
    private static readonly Regex MigrationIdPattern = new(
        "^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)+$",
        RegexOptions.CultureInvariant | RegexOptions.NonBacktracking);
    private static readonly Regex NoticeKeyPattern = new(
        "^[a-z][a-z0-9]*(?:\\.[a-z0-9_-]+)+$",
        RegexOptions.CultureInvariant | RegexOptions.NonBacktracking);

    private readonly IReadOnlyDictionary<PocketCapabilityKey, PocketCapabilityLifecycleRecord> _records;
    private readonly IReadOnlyDictionary<PocketCapabilityKey, PocketCapabilityReferenceMigration> _migrations;

    internal static PocketCapabilityCompatibilityCatalog BuiltIn { get; } = new(
        CurrentHostVersion,
        [],
        []);

    internal PocketCapabilityCompatibilityCatalog(
        string hostVersion,
        IEnumerable<PocketCapabilityLifecycleRecord> records,
        IEnumerable<PocketCapabilityReferenceMigration> migrations)
    {
        if (SemanticVersion(hostVersion) is null)
        {
            throw Failure("host_version");
        }

        var recordMap = new Dictionary<PocketCapabilityKey, PocketCapabilityLifecycleRecord>();
        foreach (var record in records)
        {
            if (!ValidKey(record.Key)
                || SemanticVersion(record.IntroducedInHostVersion) is null
                || !recordMap.TryAdd(record.Key, record))
            {
                throw Failure("record");
            }
            Validate(record, hostVersion);
        }

        var migrationMap = new Dictionary<PocketCapabilityKey, PocketCapabilityReferenceMigration>();
        var migrationIds = new HashSet<string>(StringComparer.Ordinal);
        foreach (var migration in migrations)
        {
            if (!ValidMigrationId(migration.Id)
                || !ValidKey(migration.Source)
                || !ValidKey(migration.Target)
                || migration.Source == migration.Target
                || !migrationMap.TryAdd(migration.Source, migration)
                || !migrationIds.Add(migration.Id))
            {
                throw Failure("migration");
            }
            if (!recordMap.TryGetValue(migration.Source, out var record)
                || record.Status == PocketCapabilityLifecycleStatus.Active
                || record.Replacement != migration.Target
                || record.MigrationId != migration.Id)
            {
                throw Failure("migration_binding");
            }
        }

        foreach (var record in recordMap.Values.Where(item => item.Status != PocketCapabilityLifecycleStatus.Active))
        {
            if (!migrationMap.ContainsKey(record.Key))
            {
                throw Failure("migration_missing");
            }
            if (record.Replacement is not { } replacement
                || (recordMap.TryGetValue(replacement, out var target)
                    && target.Status == PocketCapabilityLifecycleStatus.Removed))
            {
                throw Failure("replacement_removed");
            }
        }

        foreach (var source in migrationMap.Keys)
        {
            var visited = new HashSet<PocketCapabilityKey> { source };
            var current = migrationMap[source].Target;
            while (migrationMap.TryGetValue(current, out var next))
            {
                if (!visited.Add(current))
                {
                    throw Failure("migration_cycle");
                }
                current = next.Target;
            }
        }

        HostVersion = hostVersion;
        _records = recordMap;
        _migrations = migrationMap;
    }

    internal string HostVersion { get; }

    internal PocketCapabilityLifecycleStatus Status(PocketCapabilityKey key) =>
        _records.TryGetValue(key, out var record) ? record.Status : PocketCapabilityLifecycleStatus.Active;

    internal PocketCapabilityCompatibilityIssue? Issue(PocketCapabilityKey key)
    {
        if (!_records.TryGetValue(key, out var record)
            || record.Status == PocketCapabilityLifecycleStatus.Active
            || record.Replacement is not { } replacement
            || record.MigrationId is not { } migrationId
            || record.RemovalNotBeforeHostVersion is not { } removalNotBefore)
        {
            return null;
        }
        return new PocketCapabilityCompatibilityIssue(
            key,
            record.Status,
            replacement,
            migrationId,
            removalNotBefore,
            record.NoticeKey);
    }

    internal PocketCapabilityReferenceMigration Migration(PocketCapabilityKey key) =>
        _migrations.TryGetValue(key, out var migration)
            ? migration
            : throw Failure("migration_unavailable");

    internal void RequireRuntimeExecutable(PocketCapabilityKey key)
    {
        if (_records.TryGetValue(key, out var record)
            && record.Status == PocketCapabilityLifecycleStatus.Removed)
        {
            var field = record.Replacement is { } replacement
                ? $"{key.Id}@{key.Version}->{replacement.Id}@{replacement.Version}"
                : $"{key.Id}@{key.Version}";
            throw new CapabilityBrokerException("CAPABILITY_REMOVED", field);
        }
    }

    private static void Validate(PocketCapabilityLifecycleRecord record, string hostVersion)
    {
        var introduced = SemanticVersion(record.IntroducedInHostVersion);
        var host = SemanticVersion(hostVersion);
        if (!ValidNoticeKey(record.NoticeKey)
            || introduced is null
            || host is null
            || introduced > host)
        {
            throw Failure("record_version");
        }

        if (record.Status == PocketCapabilityLifecycleStatus.Active)
        {
            if (record.DeprecatedInHostVersion is not null
                || record.RemovalNotBeforeHostVersion is not null
                || record.Replacement is not null
                || record.MigrationId is not null)
            {
                throw Failure("active_fields");
            }
            return;
        }

        var deprecated = SemanticVersion(record.DeprecatedInHostVersion);
        var removal = SemanticVersion(record.RemovalNotBeforeHostVersion);
        if (deprecated is null
            || removal is null
            || record.Replacement is not { } replacement
            || !ValidKey(replacement)
            || replacement == record.Key
            || record.MigrationId is not { } migrationId
            || !ValidMigrationId(migrationId)
            || introduced > deprecated
            || deprecated > host
            || deprecated >= removal)
        {
            throw Failure("deprecation_window");
        }
        if (record.Status == PocketCapabilityLifecycleStatus.Removed && host < removal)
        {
            throw Failure("removed_too_early");
        }
    }

    private static bool ValidKey(PocketCapabilityKey key) =>
        key.Version >= 1 && CapabilityIdPattern.IsMatch(key.Id);

    private static bool ValidMigrationId(string value) =>
        value.Length <= 128 && MigrationIdPattern.IsMatch(value);

    private static bool ValidNoticeKey(string value) =>
        value.Length <= 160 && NoticeKeyPattern.IsMatch(value);

    private static Version? SemanticVersion(string? value) =>
        value is not null
        && Regex.IsMatch(value, "^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)$", RegexOptions.CultureInvariant)
        && Version.TryParse(value, out var version)
            ? version
            : null;

    private static PocketCapabilityCompatibilityException Failure(string code) => new(code);
}
